File: src/main/java/com/somefrills/features/misc/glowblock/BlockScanner.java
package com.somefrills.features.misc.glowblock;

import net.minecraft.block.Block;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.ChunkPos;
import net.minecraft.world.chunk.WorldChunk;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static com.somefrills.Main.mc;

public class BlockScanner {

    private final Set<ChunkPos> scannedChunks = new HashSet<>();
    private final Set<BlockPos> glowingBlocks = new HashSet<>();

    private ChunkPos lastCenter = null;

    public void reset() {
        clearResultsOnly();
        clearChunksOnly();
    }

    public void clearChunksOnly() {
        scannedChunks.clear();
        lastCenter = null;
    }

    public void clearResultsOnly() {
        glowingBlocks.clear();
    }

    public void removeBlockFromResults(Block block) {
        glowingBlocks.removeIf(pos ->
                mc.world.getBlockState(pos).getBlock() == block
        );
    }

    public Set<BlockPos> scanRenderedChunks(List<Block> targetBlocks) {
        if (mc.world == null || mc.player == null) {
            return glowingBlocks;
        }

        ChunkPos center = new ChunkPos(mc.player.getBlockPos());
        int radius = mc.options.getViewDistance().getValue();

        // optional: reset scan memory when player moves chunk
        if (lastCenter == null || !lastCenter.equals(center)) {
            lastCenter = center;
        }

        // scan ONLY 1 chunk per frame
        ChunkPos toScan = null;

        outer:
        for (int dx = -radius; dx <= radius; dx++) {
            for (int dz = -radius; dz <= radius; dz++) {

                ChunkPos cp = new ChunkPos(center.x + dx, center.z + dz);

                if (scannedChunks.contains(cp)) continue;
                if (!mc.world.isChunkLoaded(cp.x, cp.z)) continue;

                toScan = cp;
                break outer;
            }
        }

        if (toScan == null) return glowingBlocks;

        scannedChunks.add(toScan);

        WorldChunk chunk = mc.world.getChunk(toScan.x, toScan.z);
        BlockPos basePos = chunk.getPos().getStartPos();

        for (int x = 0; x < 16; x++) {
            for (int z = 0; z < 16; z++) {

                int worldX = basePos.getX() + x;
                int worldZ = basePos.getZ() + z;

                for (int y = chunk.getBottomY(); y <= chunk.getTopYInclusive(); y++) {

                    BlockPos pos = new BlockPos(worldX, y, worldZ);

                    if (targetBlocks.contains(mc.world.getBlockState(pos).getBlock())) {
                        glowingBlocks.add(pos);
                    }
                }
            }
        }

        return glowingBlocks;
    }
}

File: src/main/kotlin/com/somefrills/features/misc/glowblock/BlockScanner.kt
package com.somefrills.features.misc.glowblock

import com.somefrills.Main.mc
import net.minecraft.core.BlockPos
import net.minecraft.world.level.ChunkPos
import net.minecraft.world.level.block.Block

class BlockScanner {
    private val scannedChunks: MutableSet<ChunkPos> = HashSet<ChunkPos>()
    private val glowingBlocks: MutableSet<BlockPos> = HashSet<BlockPos>()

    private var lastCenter: ChunkPos? = null

    fun reset() {
        clearResultsOnly()
        clearChunksOnly()
    }

    fun clearChunksOnly() {
        scannedChunks.clear()
        lastCenter = null
    }

    fun clearResultsOnly() {
        glowingBlocks.clear()
    }

    fun removeBlockFromResults(block: Block?) {
        val world = mc.level ?: return
        glowingBlocks.removeIf { pos: BlockPos -> world.getBlockState(pos).block === block }
    }

    fun scanRenderedChunks(targetBlocks: MutableList<Block>): MutableSet<BlockPos> {
        val player = mc.player ?: return glowingBlocks
        val world = mc.level ?: return glowingBlocks

        val center = ChunkPos(player.blockX, player.blockZ)
        val radius: Int = mc.options.renderDistance().get()

        // optional: reset scan memory when player moves chunk
        if (lastCenter == null || lastCenter != center) {
            lastCenter = center
        }

        // scan ONLY 1 chunk per frame
        var toScan: ChunkPos? = null

        outer@ for (dx in -radius..radius) {
            for (dz in -radius..radius) {
                val cp = ChunkPos(center.x + dx, center.z + dz)

                if (scannedChunks.contains(cp)) continue
                if (!world.isLoaded(BlockPos(cp.x, 100, cp.z))) continue

                toScan = cp
                break@outer
            }
        }

        if (toScan == null) return glowingBlocks

        scannedChunks.add(toScan)

        val chunk = world.getChunk(toScan.x, toScan.z)
        val basePosX = chunk.getPos().minBlockX
        val basePosZ = chunk.getPos().minBlockZ

        for (x in 0..15) {
            for (z in 0..15) {
                val worldX = basePosX + x
                val worldZ = basePosZ + z

                for (y in chunk.minY..chunk.maxY) {
                    val pos = BlockPos(worldX, y, worldZ)

                    if (targetBlocks.contains(world.getBlockState(pos).block)) {
                        glowingBlocks.add(pos)
                    }
                }
            }
        }

        return glowingBlocks
    }
}